ref(metrics): Log metric outcomes via dedicated struct

Dav1dde opened the pull request:
Replaces `utils::reject_metrics` with a dedicated struct responsible to log outcomes for metrics. Also updates the rate limiter to use the new struct, which surfaced a bug where we did not have outcomes for spans in `reject_metrics`.

Follow ups:
- Always take ownership of buckets when tracking outcomes, this should be the last step in the lifecycle of a bucket
- Implement proper outcomes in the batch metric buckets callback
- Fix the metadata situation, metadata of views should (for volume) needs to be handled differently

#skip-changelog

Dav1dde committed:
wip

Dav1dde committed:
wip

Dav1dde committed:
wip

Dav1dde committed:
Merge remote-tracking branch 'origin/master' into dav1d/metric-outcomes

Dav1dde committed:
wip

Dav1dde committed:
merge conflicts and tests etc.

Dav1dde committed:
testutils thingy

jjbayer commented on relay-server/src/metrics.rs:
```diff
@@ -0,0 +1,261 @@
+pub const PROFILE_TAG: &str = "has_profile";
+
+pub enum Quantities {
+    Extract(ExtractionMode),
```
```suggestion
/// Indicates from where quantities should be taken.
pub enum Quantities {
    ExtractFromBuckets(ExtractionMode),
```

jjbayer commented on relay-server/src/services/processor.rs:
```diff
@@ -2967,21 +2962,22 @@ impl UpstreamRequest for SendMetricsRequest {
+                    for (scoping, quantities) in self.quantities {
+                        self.metric_outcomes.track(
+                            scoping,
+                            &[] as &[Bucket], // TODO(dav1d): provide buckets here
```
Is this for a follow-up?

Dav1dde replied:
Yes! We had the same problem before, it was the `None` in the `reject_metrics` call. I decided not to also add these functional changes to the PR, will do in a follow up.

jjbayer commented on relay-server/src/services/project.rs:
```diff
@@ -1126,22 +1128,12 @@ impl Project {
             if limits.is_limited() {
                 // NB: Until Vec::extract_if is stable, we have to iterate twice.
-                let matching_buckets = buckets
+                let _matching_buckets = buckets
```
Since iterators are lazy, isn't this a noop?

Dav1dde replied:
Good catch thanks, I think I messed up the merge here and I ran `clippy --fix` to fix a bunch of unused imports, lesson learned.

jjbayer commented on relay-server/src/utils/split_off.rs:
```diff
@@ -0,0 +1,39 @@
+
+        let (apples, oranges) = split_off(data, |item| {
+            if item != "orange" {
+                Either::Left(item)
+            } else {
+                Either::Right(item)
+            }
```
Personally I would find this easier to use if the predicate returned `true` or `false`.

Dav1dde replied:
A boolean doesn't allow for a different type (partition + map)

jjbayer replied:
Right. Should we introduce a dedicated enum then? Something like `SplitOff` vs. `Keep`? Not a blocker, just bikeshedding here.

Dav1dde replied:
Introduced a `split_off` which takes a boolean and delegates to the `split_off_map` now. I did like the `Either` since it matches the return value `(left, right)`.

Dav1dde committed:
rate limit fix

Dav1dde committed:
fix tests

Dav1dde committed:
rustdoc

Dav1dde committed:
fix(metric-stats): Only emit volume metadata for the first bucket view (#3609)

Splitting a bucket into multiple via a `BucketView` brings the problem
that some of the metadata can't be attributed correctly anymore
(aggregation is a lossy process for metadata). To solve this problem for
the volume metadata, it will only be attached to the bucket view
containing the start of the bucket.
